engine/src/main/java/org/terasology/math/Diamond3iIterable.java

```diff
@@ -0,0 +1,162 @@
+     * @param origin  center of diamond iterator
+     * @param start distance to start iterating from
+     * @param end distance to iterate out to
+     */
+    public static Diamond3iIterable.Builder hollow(Vector3ic origin, int start, int end) {
+        return new Diamond3iIterable.Builder(origin, end).start(start);
+    }
+
+    /**
+     * iterates in a shell that is 1 block wide
+     * @param origin center of shell
+     * @param radius distance for 1 block wide shell
+     */
+    public static Diamond3iIterable.Builder shell(Vector3ic origin, int radius) {
+        return new Diamond3iIterable.Builder(origin, radius).start(radius - 1);
+    }
+
+    @Override
+    public Iterator<Vector3ic> iterator() {
+        Vector3i pos = new Vector3i();
+        final int[] level = {this.startDistance};
+        final int[] offset = {-this.startDistance,0,0};
+
+        return new Iterator<Vector3ic>() {
+            @Override
+            public boolean hasNext() {
+                return level[0] < endDistance;
+            }
+
+            @Override
+            public Vector3ic next() {
+                pos.set(origin.x + offset[0], origin.y + offset[1], origin.z + offset[2]);
+                if (offset[2] < 0) {
+                    offset[2] *= -1;
+                } else if (offset[1] < 0) {
+                    offset[1] *= -1;
+                    offset[2] = -(level[0] - TeraMath.fastAbs(offset[0]) - TeraMath.fastAbs(offset[1]));
+                } else {
+                    offset[1] = -offset[1] + 1;
+                    if (offset[1] > 0) {
+                        if (++offset[0] <= level[0]) {
+                            offset[1] = TeraMath.fastAbs(offset[0]) - level[0];
+                            offset[2] = 0;
+                        } else {
+                            level[0]++;
+                            offset[0] = -level[0];
+                            offset[1] = 0;
+                            offset[2] = 0;
+                        }
+                    } else {
+                        offset[2] = -(level[0] - TeraMath.fastAbs(offset[0]) - TeraMath.fastAbs(offset[1]));
+                    }
+                }
+                return pos;
+            }
+        };
+    }
+
+    public static final class Builder {
+        private final Vector3ic origin;
+        private final int endDistance;
+        private int startDistance = 0;
+
+        /**
+         * Default endDistance starts at 1
+         * @param origin center region for iterator
+         * @param endDistance maximums radius
+         */
+        private Builder(Vector3ic origin, int endDistance){
+            this.origin = origin;
+            this.endDistance = endDistance + 1;
+        }
+
+        /**
+         * Default start distance is 0
+         * @param start  the minimum radius
+         */
+        public Diamond3iIterable.Builder start(int start) {
+            this.startDistance = start + 1;
```

Comment:
+/- 1 confusion:
```suggestion
            this.startDistance = start;
```
that was the last one.